File: fastqaoa/ctypes/diagonals.py
from typing import List
from numpy.ctypeslib import ndpointer
import numpy as np
import qubovert as qv
from functools import partial

from .lib import _lib, C, NP_REAL


class Diagonals(C.Structure):
    _fields_ = [
        ("n_qubits", C.c_uint8),
        ("data", C.POINTER(C.m_real)),
        ("min_val", C.m_real),
        ("max_val", C.m_real),
    ]

    LTE = 0
    GTE = 1
    LT = 2
    GT = 3
    EQ = 4
    NEQ = 5

    def make_plus_state(n_qubits: int) -> "Diagonals":
        ...

    def print(self):
        ...

    def __del__(self):
        ...

    def to_numpy(self) -> np.ndarray:
        ...

    def from_numpy(arr: np.ndarray) -> "Diagonals":
        ...

    def brute_force(
        n_qubits: int, keys: List[int], vals: List[float]  # pyright: ignore
    ) -> "Diagonals":
        ...

    def brute_force_qv(model: qv.PUBO) -> "Diagonals":
        b = {sum(1 << i for i in k) if len(k) > 0 else 0: v for k, v in model.items()}
        keys = list(b.keys())
        vals = list(b.values())

        return Diagonals.brute_force(model.num_binary_variables, keys, vals)

    def mask(
        self,
        lhs: "Diagonals",  # pyright: ignore
        rhs: float,  # pyright: ignore
        cmp_type: int,  # pyright: ignore
        val: float = 0.0,  # pyright: ignore
    ) -> "Diagonals":
        ...

    def quad_penalty(
        self,
        lhs: "Diagonals",  # pyright: ignore
        rhs: float,  # pyright: ignore
        cmp_type: int,  # pyright: ignore
        penalty: float = None,  # pyright: ignore
    ) -> "Diagonals":
        ...

    def __le__(self, other: float):  # pyright: ignore
        ...

    def __ge__(self, other: float):  # pyright: ignore
        ...

    def __lt__(self, other: float): # pyright: ignore
        ...

    def __gt__(self, other: float): # pyright: ignore
        ...

    def __eq__(self, other: float):  # pyright: ignore
        ...

    def __ne__(self, other: float):  # pyright: ignore
        ...

    def cmp(self, other: float, typ: int):  # pyright: ignore
        ...

    def clone(self) -> "Diagonals":
        ...

    def __iadd__(self, other: float) -> "Diagonals": # pyright: ignore
        ...

    def __imul__(self, other: float) -> "Diagonals": # pyright: ignore
        ...

    def __mul__(self, other: float):
        x = self.clone()
        x *= other
        return x

    def __add__(self, other: float):
        x = self.clone()
        x += other
        return x

    def __rmul__(self, other: float):
        return self.__mul__(other)

    def __radd__(self, other: float):
        return self.__add__(other)

    def __truediv__(self, other: float):
        return self.__mul__(1 / other)

_lib.dg_print.argtypes = [C.POINTER(Diagonals)]
_lib.dg_print.restype = None


def __dg_print(self):
    return _lib.dg_print(self)


Diagonals.print = __dg_print

_lib.dg_free.argtypes = [C.POINTER(Diagonals)]
_lib.dg_free.restype = None


def __dg_del(self):
    return _lib.dg_free(self)


Diagonals.__del__ = __dg_del

_lib.dg_copy.argtypes = [C.POINTER(Diagonals), ndpointer(NP_REAL)]
_lib.dg_copy.restype = None


def __to_numpy(self):
    res = np.empty(1 << self.n_qubits, dtype=NP_REAL)
    _lib.dg_copy(self, res)
    return res


Diagonals.to_numpy = __to_numpy

_lib.dg_copy_from.argtypes = [ndpointer(NP_REAL), C.c_uint8]
_lib.dg_copy_from.restype = C.POINTER(Diagonals)


def __from_numpy(arr):
    n_qubits = int(np.log2(arr.shape[0]))
    assert 1 << n_qubits == arr.shape[0], "Not a valid diagonals dimension"
    arr = arr.astype(NP_REAL)
    return _lib.dg_copy_from(arr, n_qubits).contents


Diagonals.from_numpy = __from_numpy

_lib.dg_brute_force.argtypes = [
    C.c_uint8,
    C.c_int,
    ndpointer(np.uint64),
    ndpointer(NP_REAL),
]
_lib.dg_brute_force.restype = C.POINTER(Diagonals)


def __brute_force(n_qubits, keys, vals):
    keys = np.array(keys).astype(np.uint64)
    vals = np.array(vals).astype(NP_REAL)
    assert vals.shape[0] == keys.shape[0], "Keys and vals need to match in dimension"
    return _lib.dg_brute_force(n_qubits, keys.shape[0], keys, vals).contents


Diagonals.brute_force = __brute_force

_lib.dg_mask.argtypes = [
    C.POINTER(Diagonals),
    C.POINTER(Diagonals),
    C.m_real,
    C.c_int,
    C.m_real,
]
_lib.dg_mask.restype = C.POINTER(Diagonals)


def __mask(self, lhs, rhs, typ, val=0.0):
    return _lib.dg_mask(self, lhs, rhs, typ, val).contents


Diagonals.mask = __mask


_lib.dg_quad_penalty.argtypes = [
    C.POINTER(Diagonals),
    C.POINTER(Diagonals),
    C.m_real,
    C.c_int,
    C.POINTER(C.m_real),
]
_lib.dg_quad_penalty.restype = C.POINTER(Diagonals)


def __quad_peanlty(self, lhs, rhs, typ, penalty=None):
    if penalty is None:
        penalty = C.m_real(-1)
    else:
        penalty = C.m_real(penalty)
    res = _lib.dg_quad_penalty(self, lhs, rhs, typ, penalty).contents
    return res, penalty.value


Diagonals.quad_penalty = __quad_peanlty

_lib.dg_cmp.argtypes = [C.POINTER(Diagonals), C.m_real, C.c_int]
_lib.dg_cmp.restype = C.POINTER(Diagonals)


def __cmp(lhs, rhs, typ=0):
    return _lib.dg_cmp(lhs, rhs, typ).contents

Diagonals.cmp = __cmp
Diagonals.__le__ = lambda self, other: __cmp(self, other, Diagonals.LTE)
Diagonals.__ge__ = lambda self, other: __cmp(self, other, Diagonals.GTE)
Diagonals.__lt__ = lambda self, other: __cmp(self, other, Diagonals.LT)
Diagonals.__gt__ = lambda self, other: __cmp(self, other, Diagonals.GT)
Diagonals.__eq__ = lambda self, other: __cmp(self, other, Diagonals.EQ)
Diagonals.__ne__ = lambda self, other: __cmp(self, other, Diagonals.NEQ)


_lib.dg_scale.argtypes = [C.POINTER(Diagonals), C.m_real]
_lib.dg_scale.restype = None

_lib.dg_shift.argtypes = [C.POINTER(Diagonals), C.m_real]
_lib.dg_shift.restype = None

_lib.dg_clone.argtypes = [C.POINTER(Diagonals)]
_lib.dg_clone.restype = C.POINTER(Diagonals)

def __clone(self):
    return _lib.dg_clone(self).contents
Diagonals.clone = __clone

def __mul(self, other):
    _lib.dg_scale(self, other)
    return self
Diagonals.__imul__ = __mul
Diagonals.__mul = __mul

def __add(self, other):
    _lib.dg_shift(self, other)
    return self
Diagonals.__iadd__ = __add

File: src/diagonals.c
#include "diagonals.h"
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

#define FIND_MIN_VAL(n)                                                        \
  if (n->data[i] < n->min_val)                                                 \
    n->min_val = n->data[i];
#define FIND_MAX_VAL(n)                                                        \
  if (n->data[i] > n->max_val)                                                 \
    n->max_val = n->data[i];

#define FEASIBLE                                                               \
  diff = lhs->data[i] - rhs;                                                   \
  cmp_ok = ((cmp == LTE) & (diff <= 0)) | ((cmp == GTE) & (diff >= 0)) |       \
           ((cmp == LT) & (diff < 0)) | ((cmp == GT) & (diff > 0)) |           \
           ((cmp == EQ) & (fabs(diff) <= TOL)) |                               \
           ((cmp == NEQ) & (fabs(diff) > TOL));

#define NOT_FEASIBLE                                                           \
  diff = lhs->data[i] - rhs;                                                   \
  cmp_ok = ((cmp == LTE) & !(diff <= 0)) | ((cmp == GTE) & !(diff >= 0)) |     \
           ((cmp == LT) & !(diff < 0)) | ((cmp == GT) & !(diff > 0)) |         \
           ((cmp == EQ) & !(fabs(diff) <= TOL)) |                              \
           ((cmp == NEQ) & !(fabs(diff) > TOL));

diagonals_t *dg_malloc(uint8_t n_qubits) {
  diagonals_t *dg = (diagonals_t *)malloc(sizeof(diagonals_t));
  dg->data = (real *)malloc((1 << n_qubits) * sizeof(real));
  dg->n_qubits = n_qubits;
  dg->min_val = INFINITY;
  dg->max_val = -INFINITY;
  return dg;
}

diagonals_t *dg_clone(const diagonals_t *dg) {
  diagonals_t *other = dg_malloc(dg->n_qubits);

  for (size_t i = 0; i < 1 << dg->n_qubits; i++) {
    other->data[i] = dg->data[i];
  }

  other->min_val = dg->min_val;
  other->max_val = dg->max_val;

  return other;
}

void dg_copy(const diagonals_t *dg, real *other) {
  for (size_t i = 0; i < 1 << dg->n_qubits; i++) {
    other[i] = dg->data[i];
  }
}

diagonals_t *dg_copy_from(const real *other, const uint8_t n_qubits) {
  diagonals_t *dg = dg_malloc(n_qubits);

  for (size_t i = 0; i < 1 << dg->n_qubits; i++) {
    dg->data[i] = other[i];
    FIND_MIN_VAL(dg)
    FIND_MAX_VAL(dg)
  }
  return dg;
}

void dg_free(diagonals_t *dg) {
  free(dg->data);
  free(dg);
}

void dg_print(const diagonals_t *dg) {
  for (size_t i = 0; i < 1 << dg->n_qubits; i++) {
    printf("%.3e\n", dg->data[i]);
  }
}

diagonals_t *dg_brute_force(uint8_t n_qubits, int polylen, uint64_t *polykeys,
                            real *polyvals) {
  diagonals_t *dg = dg_malloc(n_qubits);
  for (uint64_t i = 0; i < 1 << n_qubits; i++) {
    real buf = 0.;
    for (int j = 0; j < polylen; j++) {
      buf += ((i & polykeys[j]) == polykeys[j]) * polyvals[j];
    }
    dg->data[i] = buf;
    FIND_MIN_VAL(dg)
    FIND_MAX_VAL(dg)
  }
  return dg;
}

diagonals_t *dg_mask(diagonals_t *dg, diagonals_t *lhs, real rhs, cmp_kind cmp,
                     real val) {
  diagonals_t *ret = dg_malloc(dg->n_qubits);

  bool cmp_ok;
  real diff;
  for (size_t i = 0; i < 1 << dg->n_qubits; i++) {
    NOT_FEASIBLE
    if (cmp_ok) {
      ret->data[i] = val;
    } else {
      ret->data[i] = dg->data[i];
    }
    FIND_MIN_VAL(ret)
    FIND_MAX_VAL(ret)
  }

  return ret;
}

diagonals_t *dg_quad_penalty(diagonals_t *dg, diagonals_t *lhs, real rhs,
                             cmp_kind cmp, real *penalty) {
  diagonals_t *ret = dg_clone(dg);
  real diff;
  bool cmp_ok;

  // auto penalty
  if (*penalty < 0) {
    real scnd_min_val = dg->max_val + 1;
    real min_val = dg->max_val + 1;
    for (size_t i = 0; i < 1 << dg->n_qubits; i++) {
      FEASIBLE
      if (cmp_ok) {
        if (dg->data[i] < min_val) {
          scnd_min_val = min_val;
          min_val = dg->data[i];
        } else if (dg->data[i] > min_val + TOL && dg->data[i] < scnd_min_val) {
          scnd_min_val = dg->data[i];
        }
      }
    }

    if (scnd_min_val > dg->max_val)
      scnd_min_val = 0;

    *penalty = 0.;
    real new_penalty = 0;
    for (size_t i = 0; i < 1 << dg->n_qubits; i++) {
      NOT_FEASIBLE
      if (cmp_ok) {
        new_penalty = (scnd_min_val - dg->data[i]) / (diff * diff + TOL);
        if (new_penalty > *penalty) {
          *penalty = new_penalty;
        }
      }
    }
  }

  ret->min_val = INFINITY;
  ret->max_val = -INFINITY;
  for (size_t i = 0; i < 1 << dg->n_qubits; i++) {
    NOT_FEASIBLE
    if (cmp_ok) {
      ret->data[i] += *penalty * diff * diff;
    }
    FIND_MIN_VAL(ret)
    FIND_MAX_VAL(ret)
  }
  return ret;
}

diagonals_t *dg_cmp(diagonals_t *lhs, real rhs, cmp_kind cmp) {
  diagonals_t *ret = dg_malloc(lhs->n_qubits);
  ret->min_val = 0.;
  ret->max_val = 1.;

  real diff;
  bool cmp_ok;
  for (size_t i = 0; i < 1 << lhs->n_qubits; i++) {
    FEASIBLE
    ret->data[i] = (real)cmp_ok;
  }

  return ret;
}

void dg_scale(diagonals_t *diags, real f) {
  for (size_t i = 0; i < 1 << diags->n_qubits; i++) {
    diags->data[i] *= f;
  }
  if (f < 0) {
    real mv = diags->max_val;
    diags->max_val = f * diags->min_val;
    diags->min_val = f * mv;
  } else {
    diags->max_val *= f;
    diags->min_val *= f;
  }
}

void dg_shift(diagonals_t *diags, real f) {
  for (size_t i = 0; i < 1 << diags->n_qubits; i++) {
    diags->data[i] += f;
  }
  diags->min_val += f;
  diags->max_val += f;
}

real dg_expec_sample(const diagonals_t *diags, const int num,
                     const uint32_t *samples) {
  real res = 0;

  for (int i = 0; i < num; i++) {
    res += diags->data[samples[i]];
  }
  return res / (real)num;
}
